pkg/controller/deorbit/controller.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/sapcc/kubernikus/pkg/util"
+)
+
+func init() {
+	prometheus.MustRegister(
+		DeorbitOperationsLatency,
+		DeorbitOperationsTotal,
+		DeorbitSuccessfulOperationsTotal,
+		DeorbitFailedOperationsTotal,
+	)
+}
+
+const (
+	DeorbiterFinalizer = "deorbiter"
+
+	// The longest time a single worker will be blocked, while waiting for the clean
+	// up of OpenStack resources. After this timeout, deobiting will be queued again.
+	// It's a safeguard against having all workers hanging indefinetly and congesting
+	// the queue.
+	UnblockWorkerTimeout = 10 * time.Minute
+
+	// If the customer's apiserver is unreachable for this duration, we assume it is
+	// already decomissioned, permanently damaged or was never created successfully
+	// in the first place.
+	APIUnavailableTimeout = 2 * time.Minute
+
+	// After this duration, the deorbiter will utimately give up retrying. This will
+	// potentially leave debris in the customer's project in the form of volumes,
+	// load balancers or floating IPs.
+	DeorbitHangingTimeout = 24 * time.Hour
+)
+
+type DeorbitReconciler struct {
+	config.Clients
+
+	Recorder record.EventRecorder
+	Logger   log.Logger
+
+	klusterInformer informers_kubernikus.KlusterInformer
+}
+
+func NewController(threadiness int, factories config.Factories, clients config.Clients, recorder record.EventRecorder, logger log.Logger) base.Controller {
+	logger = log.With(logger, "controller", "deorbit")
+
+	var reconciler base.Reconciler
+	reconciler = &DeorbitReconciler{clients, recorder, logger, factories.Kubernikus.Kubernikus().V1().Klusters()}
+	reconciler = &base.LoggingReconciler{reconciler, logger}
+	reconciler = &base.InstrumentingReconciler{
+		reconciler,
+		metrics.DeorbitOperationsLatency,
+		metrics.DeorbitOperationsTotal,
+		metrics.DeorbitSuccessfulOperationsTotal,
+		metrics.DeorbitFailedOperationsTotal,
+	}
+	return base.NewController(threadiness, factories, clients, reconciler, logger)
+}
+
+func (d *DeorbitReconciler) Reconcile(kluster *v1.Kluster) (bool, error) {
+	switch kluster.Status.Phase {
+	case models.KlusterPhaseRunning:
+		return false, util.EnsureFinalizerCreated(d.Kubernikus.KubernikusV1(), d.klusterInformer.Lister(), kluster, DeorbiterFinalizer)
+	case models.KlusterPhaseTerminating:
+		if kluster.HasFinalizer(DeorbiterFinalizer) {
+			if err := d.deorbit(kluster); err != nil {
+				return false, err
+			}
+			return false, util.EnsureFinalizerRemoved(d.Kubernikus.KubernikusV1(), d.klusterInformer.Lister(), kluster, DeorbiterFinalizer)
+		}
+	}
+	return false, nil
+}
+
+func (d *DeorbitReconciler) deorbit(kluster *v1.Kluster) (err error) {
+	// The following channel is used to abort the deorbiting after a certain
+	// time. It is required because the deorbit wait-functions block indefinetly or
+	// until the stop channel is closed. This is used to unblock the workqueue.
+	var once sync.Once
+	done := make(chan struct{})
+	defer once.Do(func() { close(done) })
```

Comment:
Closing a channel is not necessary for the gc to collect it, so closing it here it not necessary. That means we can remove the whole `sync.Once` logic here.

Reply:
Removed the `sync` logic completely